completion_test.go

```diff
@@ -47,69 +47,139 @@ func TestCompletionUsage(t *testing.T) {
-	got1 := buf.String()
-	assert.Contains(t, got1, "ot\n")
-
-	buf.Reset()
-	Run("ot", completionFlag)
-	got2 := buf.String()
-	assert.Contains(t, got2, "ix\n")
+	cases := []struct {
+		shell       string
+		description string
+		connector   string
+	}{
+		{
+			description: "Bash shell suggestions",
+			shell:       "bash",
+			connector:   "\t",
+		},
+		{
+			description: "zsh shell suggestions",
+			shell:       "zsh",
+			connector:   ":",
+		},
+		{
+			description: "fish shell suggestions",
+			shell:       "fish",
+			connector:   "\t",
+		},
+	}
+	for _, tt := range cases {
+		t.Run(tt.description, func(t *testing.T) {
+			resetDefaultApp()
+			addTestCompletionCommands()
+
+			var buf bytes.Buffer
+			defaultApp.completionCtx.out = &buf
+			defaultApp.completionCtx.shell = tt.shell
+
+			Run("c", completionFlag, tt.shell)
+			got := buf.String()
+			log.Println(got)
+			assert.Contains(t, got, "cmd1"+tt.connector+"A cmd1 description\n")
+			assert.NotContains(t, got, "cmd2"+tt.connector+"A hidden cmd2 description")
+			assert.NotContains(t, got, "completion")
+
+			buf.Reset()
+			Run("group1", "c", completionFlag, tt.shell)
+			got = buf.String()
+			assert.Contains(t, got, "cmd1"+tt.connector+"A group1 cmd1 description\n")
+			assert.Contains(t, got, "cmd2"+tt.connector+"A group1 cmd2 description\n")
+			assert.Contains(t, got, "cmd3\n")
+
+			buf.Reset()
+			Run("group1", "cme", completionFlag, tt.shell)
+			got = buf.String()
+			assert.Zero(t, got)
+
+			buf.Reset()
+			Run("unknown", completionFlag, tt.shell)
+			got = buf.String()
+			assert.Zero(t, got)
+		})
+	}
 
-	buf.Reset()
-	Run("group3", "s", completionFlag)
-	got3 := buf.String()
-	assert.Contains(t, got3, "sub1\n")
+	noDescCases := []struct {
+		shell       string
+		description string
+		connector   string
+	}{
+		{
+			description: "powershell shell suggestions",
+			shell:       "powershell",
+			connector:   ":",
+		},
+	}
+	for _, tt := range noDescCases {
+		t.Run(tt.description, func(t *testing.T) {
+			resetDefaultApp()
+			addTestCompletionCommands()
+
+			var buf bytes.Buffer
+			defaultApp.completionCtx.out = &buf
+			defaultApp.completionCtx.shell = tt.shell
+
+			Run("c", completionFlag, tt.shell)
+			got := buf.String()
+			assert.Contains(t, got, "cmd1\n")
+			assert.NotContains(t, got, "cmd2\n")
+			assert.NotContains(t, got, "completion\n")
+
+			buf.Reset()
+			Run("group1", "c", completionFlag, tt.shell)
+			got = buf.String()
+			assert.Contains(t, got, "cmd1\n")
+			assert.Contains(t, got, "cmd2\n")
+			assert.Contains(t, got, "cmd3\n")
+
+			buf.Reset()
+			Run("group1", "cme", completionFlag, tt.shell)
+			got = buf.String()
+			assert.Zero(t, got)
+
+			buf.Reset()
+			Run("unknown", completionFlag, tt.shell)
+			got = buf.String()
+			assert.Zero(t, got)
+		})
+	}
 
-	buf.Reset()
-	Run("group3", "sub1", completionFlag)
-	got4 := buf.String()
-	assert.Contains(t, got4, "subsub1\n")
+	// TODO: how to handle unsupported shell ?? | 2023-08-06
```

Comment:
For unsupported shells, we can fallback to generate completion without description, that should be the basic ability for a shell completion.

Reply:
Ok, we can. It would be somewhat safe.